cpp/src/sampling/detail/prepare_next_frontier_impl.cuh

```diff
@@ -0,0 +1,195 @@
+/*
+ * Copyright (c) 2022-2023, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cugraph/algorithms.hpp>
+#include <cugraph/detail/shuffle_wrappers.hpp>
+#include <cugraph/graph.hpp>
+#include <cugraph/graph_view.hpp>
+#include <cugraph/utilities/thrust_tuple_utils.hpp>
+#include <cugraph/vertex_partition_device_view.cuh>
+
+#include <sampling/detail/sampling_utils.hpp>
+
+#include <raft/core/handle.hpp>
+
+#include <thrust/optional.h>
+#include <thrust/sort.h>
+#include <thrust/tuple.h>
+#include <thrust/unique.h>
+
+#include <rmm/device_uvector.hpp>
+
+namespace cugraph {
+namespace detail {
+
+template <typename vertex_t, typename label_t, bool multi_gpu>
+std::tuple<rmm::device_uvector<vertex_t>,
+           std::optional<rmm::device_uvector<label_t>>,
+           std::optional<std::tuple<rmm::device_uvector<vertex_t>,
+                                    std::optional<rmm::device_uvector<label_t>>>>>
+prepare_next_frontier(
+  raft::handle_t const& handle,
+  raft::device_span<vertex_t const> sampled_src_vertices,
+  std::optional<raft::device_span<label_t const>> sampled_src_vertex_labels,
+  raft::device_span<vertex_t const> sampled_dst_vertices,
+  std::optional<raft::device_span<label_t const>> sampled_dst_vertex_labels,
+  std::optional<std::tuple<rmm::device_uvector<vertex_t>,
+                           std::optional<rmm::device_uvector<label_t>>>>&& vertex_used_as_source,
+  vertex_partition_view_t<vertex_t, multi_gpu> vertex_partition,
+  std::vector<vertex_t> const& vertex_partition_range_lasts,
+  prior_sources_behavior_t prior_sources_behavior,
+  bool dedupe_sources,
+  bool do_expensive_check)
+{
+  vertex_partition_device_view_t<vertex_t, multi_gpu> d_vertex_partition(vertex_partition);
+
+  size_t frontier_size = sampled_dst_vertices.size();
+  if (prior_sources_behavior == prior_sources_behavior_t::CARRY_OVER) {
+    frontier_size += sampled_src_vertices.size();
+  }
+
+  rmm::device_uvector<vertex_t> frontier_vertices(frontier_size, handle.get_stream());
+  auto frontier_vertex_labels =
+    sampled_dst_vertex_labels
+      ? std::make_optional<rmm::device_uvector<label_t>>(frontier_size, handle.get_stream())
+      : std::nullopt;
+
+  raft::copy(frontier_vertices.begin(),
+             sampled_dst_vertices.data(),
+             sampled_dst_vertices.size(),
+             handle.get_stream());
```

Comment:
Copy and paste from a `.cpp` file, that's all.  Probably better to use `thrust::copy` in this context.

Reply:
Yeah.. I feel  like `raft::update_host` and `raft::update_device` make sense (as they better document the intention), but not sure about `raft::copy` (yes, it can be used in .cpp but besides that, I am not sure. And if we are copying between device arrays, we might be calling other thrust functions as well, so we need .cu anyways).